### PR TITLE
Feature/leaderboard enhancements

### PR DESCRIPTION
**Pull request by unusedusername01**
# ✨ Summary

This PR implements a comprehensive **leaderboard position tracking and display system** across the application. Users can now see their ranking position in the leaderboard based on kudos count, with positions displayed both in the leaderboard cards and as a new stat on user profiles. The implementation includes visual enhancements with colored position badges, improved UI consistency through labeled sort options, and dynamic range filters that adapt to actual data.
Addresses issue:
- #305 
- #347 
And user story:
- #350

**Key improvements:**
- Leaderboard positions are calculated consistently using kudos-based ranking with tie-breaking logic
- Profile pages now display user's Helper's Rank with color-coded badges for podium positions
- Leaderboard cards show position numbers with underlined styling and theme-appropriate colors
- Sort options use human-readable labels instead of enum names
- Range filters dynamically adjust maximum bounds based on actual dataset values

---

# 📦 What's Included

### Core Implementation
- **`LeaderboardPositionCalculator`**: New utility class that calculates user positions based on kudos ranking with consistent tie-breaking (kudos → helpReceived → id)
- **Position Display System**: Positions shown on leaderboard cards with `#` prefix and underlined styling
- **Profile Stats Enhancement**: Added `leaderboardPosition` to `ProfileState` with dedicated position card in stats section
- **Color-Coded Badges**: Podium positions (1st-3rd) display in gold/silver/bronze colors matching badge themes

### State Management / Logic Improvements
- Added `positions: Map<String, Int>` to `LeaderboardState` for centralized position tracking
- Updated `LeaderboardViewModel` to calculate and maintain positions alongside profiles
- Enhanced `ProfileViewModel` with `calculateLeaderboardPosition()` supporting offline mode via cache fallback
- Modified `KUDOS_DESC` comparator to use helpReceived as secondary sort (consistency with position calculation)
- Removed `profileId` field from `ProfileState` and related UI components (privacy/simplification)

### UI/UX Enhancements
- **Leaderboard Cards**: 
  - Split name into two lines (first name + last name) for better readability
  - Position text displayed above medals for podium ranks, centered vertically for non-podium
  - Reduced card inner padding (12dp → 8dp) and row spacing (8dp → 4dp) for tighter layout
  - Section labels now support two-line display with ellipsis
- **Profile Actions**: Updated icons to `FormatListBulleted` and `PlaylistAddCheck` for better semantic meaning
- **Badge Themes**: Unified medal icons to `Icons.Filled.Stars` for gold/silver/bronze
- **Sort Labels**: Human-readable labels like "Kudos (Descending)" instead of "KUDOS_DESC"

### Code Quality Enhancements
- Introduced `ConstantLeaderboard.TwoLineMax` for consistent multi-line text handling
- Added `isPodiumTheme()` utility function to `LeaderboardBadgeThemes`
- Consolidated request sort definitions with explicit `label` parameter
- Reduced default range filter maximums from 500 → 50 for more realistic defaults
- Removed unused status-based sorting from `RequestSort` enum
- Test improvements: Added `RequestSort` label references, profile picture display assertions

---

# 🧪 Testing Implementation

### Test Infrastructure
- **New Test Tag**: `LeaderboardTestTags.CARD_POSITION` for position text verification
- **Profile Test Tag**: `ProfileTestTags.PROFILE_STAT_TOP_RANK` for rank stat testing
- Updated tests to use `RequestSort.TITLE_ASCENDING.label` and `RequestSort.MOST_PARTICIPANTS.label` for sort validation
- Added `assertIsDisplayed()` before logout action in `ProfileLogoutTests` for stability

### State / Logic Testing
- Removed obsolete `profileId` assertions from `ProfileUiTests` (3 test methods updated)
- Sort label tests now verify against enum-defined labels ensuring consistency
- Test coverage maintained across state changes, UI rendering, and user interactions

### Edge Cases Handled
- Empty profile lists return empty position maps
- Offline mode fallback for position calculation uses cached profiles
- Position calculation failures gracefully return null (position not displayed)
- Range filter bounds automatically adjust when no data available

---

# 📸 Screenshots / Visuals

<table width="100%">
  <tr>
    <td width="50%" valign="top">
      <h3>Updated Leaderboard View</h3>
      
      <p>
        Features position numbers with underlined styling, color-coded badges
        for top 3, split name display, and tighter card layout.
      </p>
    </td>
    <td width="50%" valign="top">
      <h3>Updated Profile Stats</h3>
      
      <p>
        Shows new "Helper's Rank" card with color-coded position display
        alongside existing kudos/help and followers/following stats.
      </p>
    </td>
  </tr>
</table>

---

# 🚀 What's Not Included (Future PRs)

- Real-time position updates when kudos change
- Position change indicators (↑↓ compared to previous period)
- Leaderboard history tracking
- Position-based notifications or achievements
- Advanced filtering by position ranges
- Animated transitions when positions change
- Detailed position breakdown by section/department

---

# 🔧 Technical Improvements

| Area | Improvement | Benefit |
|------|--------------|----------|
| Position Calculation | Centralized `LeaderboardPositionCalculator` with consistent tie-breaking | Single source of truth prevents ranking inconsistencies |
| State Management | Added `positions` map to `LeaderboardState` | Separates position calculation from display logic |
| Offline Support | Position calculation falls back to cached profiles | Works seamlessly without network connection |
| Dynamic Filtering | Range facets adjust `maxBound` based on actual data | Sliders automatically scale to dataset |
| Code Consistency | Sort enums now include explicit labels | Eliminates string manipulation, improves i18n readiness |
| UI Efficiency | Reduced padding/spacing constants | Fits more content while maintaining visual clarity |
| Testing | Label-based test assertions instead of hardcoded strings | Tests adapt to label changes automatically |

---

# 📝 Notes

**Design Decisions:**
- Position calculation uses the same comparator as `KUDOS_DESC` sort to ensure what users see matches their actual ranking
- ProfileId was removed from the UI as it provided minimal user value and cluttered the information section
- Medal icons were unified to Stars for visual consistency across all podium positions
- Two-tier position display strategy: medals with position text for podium ranks, centered text for others

**Performance Considerations:**
- Position calculation happens once per profile load/refresh, not on every render
- Position map stored in state prevents recalculation during UI updates
- Offline mode leverages existing cache infrastructure for zero network overhead

**Trade-offs:**
- Position calculation requires loading all profiles, which may not scale to very large user bases (consider pagination in future)
- Current implementation doesn't handle real-time position changes (acceptable for MVP)
- Range filter dynamic bounds require full dataset scan (negligible for current data volumes)

**Dependencies:**
- Maintains backward compatibility with existing leaderboard and profile features
- No new external dependencies introduced
- Leverages existing `UserProfileCache` for offline support

---

**Checklist**
- [x] Feature implementation complete  
- [x] Unit/UI tests added and passing  
- [x] Code reviewed for maintainability and readability  
- [x] Documentation / comments updated  

---

**Type:** `feature/leaderboard-enhancements`  
**Scope:** Leaderboard & Profile  
**Issues References:**
- #305 
- #347

> [!NOTE]
> _This PR description as well as code introduced by this PR was made with the help of LLMs_

### CODE REVIEW

File: app/src/main/java/com/android/sample/ui/profile/composables/ProfileStats.kt [hunk 1]
```diff
@@ -103,6 +105,42 @@ fun StatGroupCard(
...
+        Column(
+            modifier =
+                Modifier.fillMaxSize()
+                    .padding(
+                        vertical = ProfileDimens.StatCardVerticalPadding,
+                        horizontal = ProfileDimens.StatCardHorizontalPadding),
+            horizontalAlignment = Alignment.CenterHorizontally,
+            verticalArrangement = Arrangement.Center) {
+              Text(
+                  text = "Helper's\nRank",
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/composables/ProfileStats.kt]: Do you know what insanity is? It's keeping on putting magic values in your code, over and over again
image

File: app/src/main/java/com/android/sample/ui/request/RequestSearchFilterViewModel.kt [hunk 2]
```diff
@@ -208,36 +208,34 @@ class RequestSearchFilterViewModel(
...
           .thenBy { it.startTimeStamp }
-          .thenBy { it.requestId }),
+          .thenBy { it.requestId },
+      "Title (A-Z)"),
   TITLE_DESCENDING(
       compareByDescending<Request> { it.title.lowercase() }
           .thenBy { it.startTimeStamp }
-          .thenBy { it.requestId });
+          .thenBy { it.requestId },
+      "Title (Z-A)");
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/request/RequestSearchFilterViewModel.kt]: MAGIC VALUE ALERT

File: app/src/main/java/com/android/sample/ui/leaderboard/LeaderboardPositionCalculator.kt
[the remark this answers is not in the document]
Comment (unusedusername01) [reply app/src/main/java/com/android/sample/ui/leaderboard/LeaderboardPositionCalculator.kt]: 😊
